features/terrain.py

```diff
@@ -0,0 +1,234 @@
+from lettuce import *
+import os.path
+import psycopg2
+import requests
+import subprocess
+import shutil
+import tempfile
+import time
+import yaml
+
+
+ETCD_VERSION_URL = 'http://127.0.0.1:2379/version'
+ETCD_CLEANUP_URL = 'http://127.0.0.1:2379/v2/keys/service/batman?recursive=true'
+PATRONI_CONFIG = '{}.yml'
+etcd_handle = None
+etcd_dir = None
+pctl = None
+
+
+@world.absorb
+class PatroniController(object):
+    """ starts and stops individual patronis"""
+
+    def __init__(self):
+        self.processes = {}
+        self._patroni_path = None
+        self.connstring = {}
+        self.connections = {}
+        self.cursors = {}
+        self.log = {}
+        self.config = {}
+        self.availability_check_time_limit = 10
+        self.output_dir = None
+
+    @property
+    def patroni_path(self):
+        if self._patroni_path is None:
+            cwd = os.path.realpath(__file__)
+            while True:
+                path, entry = os.path.split(cwd)
+                cwd = path
+                if entry == 'features' or cwd == '/':
+                    break
+            self._patroni_path = cwd
+        return self._patroni_path
+
+    def patroni_is_running(self, pg_name):
+        return pg_name in self.processes and self.processes[pg_name].pid and (self.processes[pg_name].poll() is None)
+
+    def stop_patroni(self, pg_name, kill=False):
+        while self.patroni_is_running(pg_name):
+            if not kill:
+                self.processes[pg_name].terminate()
+            else:
+                self.processes[pg_name].kill()
+            time.sleep(1)
+        self.log.get('pg_name') and self.log[pg_name].close()
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Expression "(self.log.get('pg_name')) and (self.log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2110318102/issues/source?bid=3127594&fileBranchId=3144623#l57)